reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+                )
+        except (KeyError, AttributeError):
+            raise ParameterError(
+                'Invalid ref: "{}" on repo: "{}"'.format(ref, repo_url)
+            )
+
+
+def diff_missing(
+    a: Mapping[str, TFRepo],
+    b: Mapping[str, TFRepo],
+    requireDelete: bool = False,
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that are present in mapping a but missing in mapping b
+    When requireDelete = True, each repo from a is checked to ensure that the delete flag is set"""
+    missing: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key)
+        if b_repo is None:
+            if requireDelete is True and a_repo.delete is not True:
+                raise ParameterError(
+                    'To delete the terraform repo "{}", you must set delete: true in the repo definition'.format(
+                        a_repo.name
+                    )
+                )
+            missing[a_key] = a_repo
+
+    return missing
+
+
+def diff_changed(
+    a: Mapping[str, TFRepo], b: Mapping[str, TFRepo]
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that have changed between a and b in the form of returning their b values
+    In order for Terraform to work as expected, the tenant can only update the ref between MRs
+    Updating account, repo, or project_path can lead to unexpected behavior so we error
+    on that
+    """
+    changed: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key, a_repo)
+        if (
+            a_repo.account != b_repo.account
+            or a_repo.name != b_repo.name
+            or a_repo.project_path != b_repo.project_path
+            or a_repo.repository != b_repo.repository
+        ):
+            raise ParameterError(
+                'Only the `ref` and `delete` parameters for a terraform repo may be updated in merge requests on repo: "{}"'.format(
+                    a_repo.name
+                )
+            )
+        if (a_repo.ref != b_repo.ref) or (a_repo.delete != b_repo.delete):
+            changed[a_key] = b_repo
+
+    return changed
+
+
+def gql_to_tf_repo(repo: TerraformRepoV1) -> TFRepo:
+    """
+    Converts the GQL/state representation of a Terraform repo to a
+    TFRepo class that can be more easily marshalled/unmarshalled from state file
+    """
+    return TFRepo(
+        name=repo.name,
+        repository=repo.repository,
+        ref=repo.ref,
+        project_path=repo.project_path,
+        account=AWSAccount(
+            name=repo.account.name,
+            uid=repo.account.uid,
+            secret=AWSAuthSecret(
+                path=repo.account.automation_token.path,
+                version=repo.account.automation_token.version,
+            ),
+        ),
+        delete=repo.delete,
+    )
+
+
+def merge_results(
+    created: Mapping[str, TFRepo], updated: Mapping[str, TFRepo]
```

Comment:
I would use `created: Iterable[TFRepo], updated: Iterable[TFRepo]` instead, then the methods would be much easier to read.